### ISSUE TITLE
Product Setup for sequence

### ISSUE THREAD

**Issue by bash99** [chrislusf/seaweedfs#1939]
Hello, I've noticed that "seaweedfs/weed/sequence/" has memory_sequencer.go and etcd_sequencer.go, and memory_sequencer.go has say it's only for test?

But I've not found any document for use etcd as sequence storage? So how to use it?
If we only use a simple architect with volume/master, no filer, does we still need etcd, or a simple snowflow like fid is enough?

Actually we are use a old version 1.43 and got a strange bug. version 1.43 don't have etcd_sequencer.go so memory_sequencer is use in production.

After 2 years stable in production, in yesterday, clients had got same fid just with different cookie from master's /dir/assign api, not every time but very likely, as the fid is very small, 000, 02b or sometimes very big, 0xfffffffff .

So it's should be some volume server send a wrong fid and made master's fid overflow and start from 0.

But in last code, it seems master still use heartbeat from volume?
master_grpc_server.go: 64
ms.Topo.Sequence.SetMax(heartbeat.MaxFileKey)
and etcd_sequencer.go: 110
instead of collecting the max value from volume server,
the max value should be saved in local config file and ETCD cluster

I'm confused.

We had to make a hot fix on 1.43 version, use a timestamp prefix sequence to solve this before found the wrong volume. 

And we also consider upgrade to last version, but where can we find a solid config for production? I don't think just one volume server error should made the whole system wrong.

**Comment by chrislusf** [comment]
do you have multiple masters?

**Comment by bash99** [comment]
> do you have multiple masters?

Yes, 3 master run as cluster. but the seq/fid is reset to 0 even without master restart or leader re-elected.

**Comment by chrislusf** [comment]
I guess one of the master is not connected to the other 2 masters.

**Comment by bash99** [comment]
> I guess one of the master is not connected to the other 2 masters.

No, /cluster/status is OK on all masters.
We have found some volume has very big MaxFileKey like 18446744073700427373.

Can weed shell's volume fix command fix them?

**Comment by craveica** [comment]
@chrislusf I've noticed that weed scaffold -config master generates master.sequencer as type raft now and not memory as before. How does this affects when loading the old master.toml with type memory?

**Comment by chrislusf** [comment]
> We have found some volume has very big MaxFileKey like 18446744073700427373.
> 
> Can weed shell's volume fix command fix them?

Hard to fix it unless have some special code to skip these large MaxFileKey when `weed fix`.

**Comment by chrislusf** [comment]
> @chrislusf I've noticed that weed scaffold -config master generates master.sequencer as type raft now and not memory as before. How does this affects when loading the old master.toml with type memory?

`raft` is the same as `memory`

**Comment by bash99** [comment]
> > We have found some volume has very big MaxFileKey like 18446744073700427373.
> > Can weed shell's volume fix command fix them?
> 
> Hard to fix it unless have some special code to skip these large MaxFileKey when `weed fix`.

So a snowflake solution would be more robust than use maxFileKey from volume server?

**Comment by chrislusf** [comment]
> So a snowflake solution would be more robust than use maxFileKey from volume server?

yes, but harder to compress in memory.

**Comment by chrislusf** [comment]
We can experiment with snowflake file key generation. 

@bash99  Do you want to send a PR?

**Comment by bash99** [comment]
> We can experiment with snowflake file key generation.
> 
> @bash99 Do you want to send a PR?

I'll make a try, now it's a dirty solution.
I'm a noob in golang.

BTW, where is those code related to compress in memory?

**Comment by chrislusf** [comment]
> BTW, where is those code related to compress in memory?

https://github.com/chrislusf/seaweedfs/blob/master/weed/storage/needle_map/compact_map.go

**Comment by bash99** [comment]
> We can experiment with snowflake file key generation.
> 
> @bash99 Do you want to send a PR?

https://github.com/chrislusf/seaweedfs/pull/1945, any comments is welcome

**Comment by kmlebedev** [comment]
> > do you have multiple masters?
> 
> Yes, 3 master run as cluster. but the seq/fid is reset to 0 even without master restart or leader re-elected.

may be enough to enable the parameter resumeState=true ? starting from versions 2.03
https://github.com/chrislusf/seaweedfs/pull/1505/files

**Comment by bash99** [comment]
> > > do you have multiple masters?
> > 
> > 
> > Yes, 3 master run as cluster. but the seq/fid is reset to 0 even without master restart or leader re-elected.
> 
> @bash99
> may be enough to enable the parameter resumeState=true ? starting from versions 2.03
> #1503

But resumeState will still accept maxkey from volume server?
Our problem is some volume server going crazy (maybe memory or disk error) and send a 0xfffffffffffffffe maxkey to master, while master accept it, so sequence overflowed to 0.

**Comment by bash99** [comment]
> > BTW, where is those code related to compress in memory?
> 
> https://github.com/chrislusf/seaweedfs/blob/master/weed/storage/needle_map/compact_map.go

As far as I can understand, it's use delta to compress, batch size 10000. In those 10000 fids, if max(fid) - min(fid) < 2^32, we can have a good compress chance?

I'm thinking about use some twisted snowflake algorithm, for example
```
+---------------------------------------------------------------------------+
| 16 Bit NodeID  | 32 Bit Epoch Secs | 16 Bit Sequence ID |
+---------------------------------------------------------------------------+
```
instead of current default
```
+--------------------------------------------------------------------------+
| 1 Bit Unused | 41 Bit Timestamp |  10 Bit NodeID  |   12 Bit Sequence ID |
+--------------------------------------------------------------------------+
```

The new approach can support 65536 id per seconds, and if all 10000 id is generate in 65536 seconds (18.2 hours), it's will be in a 2^32 range.

But it'll lost time-ordered feature of snowflake if master leader changed.

**Comment by chrislusf** [comment]
we can use the default snowflake implementation. Just need to use `-index=leveldb` when starting `weed volume`

**Comment by bash99** [comment]
> we can use the default snowflake implementation. Just need to use `-index=leveldb` when starting `weed volume`

oh, what's will happens if we mix leveldb-index and memory-index volume?
I think if it's a fresh volume-server, it's should be fine with old volume run in memory mode.
But if some old volume is memory-index, can we restart volume server with -index=leveldb options?

**Comment by chrislusf** [comment]
> But if some old volume is memory-index, can we restart volume server with -index=leveldb options?

yes.
